### PR TITLE
Implement Tracker Commands

### PR DESCRIPTION
**Pull request by debbiextan**
Tracker Commands include:
- StartTrackerCommand
- PauseTrackerCommand
- StopTrackerCommand
- ViewTrackerCommand

Tracker-related classes include:
- Tracker
- TrackerList

Tracker-related file: tracker.txt

### CODE REVIEW

**Review by jasperosy — COMMENTED** [review]
DeveloperGuide:
I have already included the capability to automatically assign figure caption number for each diagram, please adhere to the format and apply an anchor and figure label to the diagrams you have added.
Please don't submit an implementation that is half done. Consider a top-down approach by thinking of possible alternatives before you start coding. In that case, your rationale and alternatives considered section would not be blank.

UserGuide:
Try not to just create subfolders to store your images as I am not sure if the bot will pick them up. 
Remember to include an anchor and figure label for each diagram.
Please put in some effort to at least state an example usage of your feature. e.g when a user enters the command.... this happens...

Ui:
```suggestion
        for (int i = 0; i < trackerList.getSize(); i++) {
            oof.model.tracker.Tracker tracker = trackerList.getTracker(i);
            int timeTaken = (int) tracker.getTimeTaken();
            timeTaken /= 10;
```
Magic number in method. Even if its just division, please insert a meaningful const. 
There is a dirty typecast to tracker.getTimeTaken(), please rectify.

Output.txt:
Why is there an entry with a /add description?
```Suggestion
T||N||/add buy chocolates||28-10-2019||||||
```
Is that just for testing or are you using doing some string parsing?

ViewTrackerCommand:
```Suggestion
    public void execute(SemesterList semesterList, TaskList tasks, TrackerList trackerList, Ui ui, Storage storage)
            throws OofException {
        trackerList = timeSpentByModule(tasks);
        if (trackerList.getSize() == 0) {
            throw new OofException("There are no completed Tasks!");
        }
```
Magic number, try using isEmpty() or set a constant

Storage:
```Suggestion
    private Tracker processLine(String line) {
        String[] processed = line.split("\t");
        String module = processed[0];
        String startDate = processed[1];
        String lastUpdated = processed[2];
        long timeTaken = Long.parseLong(processed[3]);
        return new Tracker(module, startDate, lastUpdated, timeTaken);
    }
```
More magic numbers, please rectify

File: docs/UserGuide.adoc [hunk 1]
```diff
@@ -360,6 +366,64 @@ Opens menu to view, add, edit and remove Semester and Module data.
...
+
+*Options for TIME_PERIOD* +
+[horizontal]
+Day:: filter time spent on each `Module` today
+Week:: filter time spent on each `Module` over the course of the last 7 days
+
+Example: `viewTracker Day`
+
+image::demo_images/Demo_ViewTrackerTimePeriod.png[]
+
```
Comment (kahlokkee) [inline docs/UserGuide.adoc]: Why is the task tracker tracking module when it is a TASK tracker and module is not a TASK. This feature is meant to track individual tasks such that the user can know how much time they spent on each task. Tracking time spent on a module does not provide any insightful details to the user. There is a reason why each module contains a list of assignments. "Time spent" belongs to the task (assignment) and when you navigate to a module, you can navigate to each individual assignment and view the time taken for that assignment.

Also, there should not be 4 different commands for the same implementation as it will cause confusion for the user and it does not fit with the command style that we use.

Lastly, the UG is missing the source files for your images. Is it intentional for the UG to be merged while incomplete?

File: src/main/java/oof/CommandParser.java [hunk 2]
```diff
@@ -95,8 +97,14 @@ public static Command parse(String line) throws OofException {
...
             line = line.replaceFirst("threshold", "").trim();
             return new ThresholdCommand(line);
+        case "pause":
+            line = line.replaceFirst("pause", "").trim();
+            return new PauseTrackerCommand(line);
         case "semester":
             return new SemesterCommand();
+        case "viewTracker":
+            line = line.replaceFirst("viewTracker", "").trim();
+            return new ViewTrackerCommand(line);
```
Comment (kahlokkee) [inline src/main/java/oof/CommandParser.java]: Refrain from using camelcase for your commands as it is inconsistent with our past practices

File: src/main/java/oof/Storage.java [hunk 3]
```diff
@@ -161,6 +165,61 @@ public void writeTaskList(TaskList taskList) {
...
+     * Reads Tracker objects that were previously saved to hard disk.
+     * @return TrackerList that contains Tracker objects.
+     * @throws OofException if unable to read tracker.txt.
+     */
+    public ArrayList<Tracker> readTrackerList() throws OofException {
+        try {
+            File file = new File(PATH_TRACKER);
+            FileReader fileReader = new FileReader(file);
+            BufferedReader br = new BufferedReader(fileReader);
+            ArrayList<oof.model.tracker.Tracker> trackerList = new ArrayList<>();
```
Comment (kahlokkee) [inline src/main/java/oof/Storage.java]: do not use such conventions.

File: src/main/java/oof/Ui.java [hunk 4]
```diff
@@ -733,7 +734,7 @@ public void printDeadlineAndEvent(String taskTime, String taskName) {
      *
      * @param taskName Name of todo.
      */
-    public void printTodo(String taskName) {
+    private void printTodo(String taskName) {
         if (taskName.length() > TEXT_SIZE_LONG) {
             taskName = taskName.substring(DESCRIPTION_LONG_START, DESCRIPTION_LONG_END);
             System.out.print(" " + taskName + ".. |");
```
Comment (kahlokkee) [inline src/main/java/oof/Ui.java]: Why are these even changed? Revert these changes.

File: src/main/java/oof/command/Command.java [hunk 5]
```diff
@@ -28,12 +29,14 @@
...
      * @param tasks        Instance of TaskList that stores Task objects.
+     * @param trackerList  Instance of TrackerList that stores Tracker objects.
      * @param ui           Instance of Ui that is responsible for visual feedback.
      * @param storage      Instance of Storage that enables the reading and writing of Task
      *                     objects to hard disk.
      * @throws OofException Catches invalid commands given by user.
      */
-    public abstract void execute(SemesterList semesterList, TaskList tasks, Ui ui, Storage storage) throws OofException;
+    public abstract void execute(SemesterList semesterList, TaskList tasks, TrackerList trackerList,
+                                 Ui ui, Storage storage) throws OofException;
```
Comment (kahlokkee) [inline src/main/java/oof/command/Command.java]: Why is TrackerList added to the argument parameter when it is not used by most commands.

**Review by kahlokkee — CHANGES_REQUESTED** [review]
Stop going around auto optimizing everything that you see.

File: docs/UserGuide.adoc (hunk 1 above)
Comment (debbiextan) [reply docs/UserGuide.adoc]: I have changed the implementation of the Tracker feature as suggested. UG has also been corrected.

File: src/main/java/oof/CommandParser.java (hunk 2 above)
Comment (debbiextan) [reply src/main/java/oof/CommandParser.java]: This issue has been rectified as suggested.

File: src/main/java/oof/Storage.java (hunk 3 above)
Comment (debbiextan) [reply src/main/java/oof/Storage.java]: This issue has been rectified as suggested.

File: src/main/java/oof/Ui.java (hunk 4 above)
Comment (debbiextan) [reply src/main/java/oof/Ui.java]: Changes have now been reverted.

File: src/main/java/oof/command/Command.java (hunk 5 above)
Comment (debbiextan) [reply src/main/java/oof/command/Command.java]: This has now been rectified.

**Comment by debbiextan** [comment]
Code changes for v1.3 are done.
UG tone edited to be more friendly.

Do take a look.

File: src/main/java/oof/command/StartTrackerCommand.java [hunk 6]
```diff
@@ -2,88 +2,113 @@
...
+                }
+                if (!isFound) {
+                    throw new OofException("Invalid Assignment description!");
+                }
+            } else {
+                updateTrackerList(trackerList);
+                storage.writeTrackerList(trackerList);
+                ui.printStartAtCurrent(tracker);
+            }
         }
```
Comment (kahlokkee) [inline src/main/java/oof/command/StartTrackerCommand.java]: This method of implementation does not take into account the possibility of multiple assignments having the same name (e.g. "assignment 1" for different modules). If there are multiple assignments with the same name, some of them are unable to be tracked at all.

File: src/main/java/oof/model/tracker/Tracker.java [hunk 7]
```diff
@@ -0,0 +1,103 @@
+package oof.model.tracker;
+
+import oof.model.task.Assignment;
+
+import java.text.DecimalFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+
+public class Tracker extends Assignment {
```
Comment (kahlokkee) [inline src/main/java/oof/model/tracker/Tracker.java]: Why is tracker inheriting from assignment when it is not a "task" per se.

File: src/main/java/oof/Storage.java [hunk 8]
```diff
@@ -177,6 +192,78 @@ public void writeTaskList(TaskList taskList) {
...
+
+        if (startDate.equals("null")) {
+            start = null;
+        } else {
+            start = readFormat.parse(startDate);
+        }
+        Date updated = readFormat.parse(lastUpdated);
+
+        return new Tracker(moduleCode, description, deadline, start, updated, timeTaken);
+    }
```
Comment (kahlokkee) [inline src/main/java/oof/Storage.java]: Status of the tracker is not being processed so it is always set to not completed even if it is saved as completed

File: src/main/java/oof/command/StopTrackerCommand.java [hunk 9]
```diff
@@ -26,52 +25,48 @@
...
-            throw new OofException("Task has already been completed.");
-        } else if (!isStarted(task)) {
-            throw new OofException("Task has no Start time.");
+
+        TrackerList trackerList = storage.readTrackerList();
+        Tracker tracker = trackerList.findTrackerByDesc(description);
+        boolean isCompleted = tracker.getStatus();
+
+        if (isCompleted) {
+            throw new OofException("Assignment has already been completed.");
```
Comment (kahlokkee) [inline src/main/java/oof/command/StopTrackerCommand.java:67]: This condition is never true because of your Storage bugs.

File: src/main/java/oof/command/ViewTrackerCommand.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+                }
+                if (!isFound) {
+                    long timeTaken = tracker.getTimeTaken();
+                    ModuleTracker moduleTracker = new ModuleTracker(moduleCode, timeTaken);
+                    moduleTrackerList.addModuleTracker(moduleTracker);
+                }
+            }
+        }
+        return moduleTrackerList;
+    }
```
Comment (kahlokkee) [inline src/main/java/oof/command/ViewTrackerCommand.java]: Arrow head.

**Review by kahlokkee — CHANGES_REQUESTED** [review]
Expect more rigorous testing especially for edge cases. I can think of multiple scenarios where your entire tracker will break permanently. The implementation of Tracker and its related classes are also a little weird. Task tracking is supposed to be a feature extension for Assignments but your implementation is totally independent of Assignment. For example, marking an Assignment as done using DoneCommand does not interact with the Tracker class at all and the Tracker are unable to see whether the Assignment is marked as done (Why does tracker even have its own "done" status??).

You'll have to seriously reconsider the feasibility and practicality of your implementation and how it should be integrated to the existing features rather than creating a totally independent feature.

File: src/main/java/oof/command/StartTrackerCommand.java (hunk 6 above)
Comment (debbiextan) [reply src/main/java/oof/command/StartTrackerCommand.java]: The code has been rectified as suggested.
Start, Stop and Pause commands now take in both module code and description as parameters.

File: src/main/java/oof/model/tracker/Tracker.java (hunk 7 above)
Comment (debbiextan) [reply src/main/java/oof/model/tracker/Tracker.java]: The code has been rectified as suggested.
Tracker now makes reference to Assignment Tasks when necessary instead.

File: src/main/java/oof/Storage.java (hunk 8 above)
Comment (debbiextan) [reply src/main/java/oof/Storage.java]: The code has been rectified as suggested.
The tracker now updates the TaskList's Task variable (Status property) correctly.

File: src/main/java/oof/command/StopTrackerCommand.java (hunk 9 above)
Comment (debbiextan) [reply src/main/java/oof/command/StopTrackerCommand.java:67]: The code is now refactored.

File: src/main/java/oof/command/ViewTrackerCommand.java (hunk 10 above)
Comment (debbiextan) [reply src/main/java/oof/command/ViewTrackerCommand.java]: The code has been rectified and now has methods where appropriate.

**Review by jasperosy — COMMENTED** [review]
Manual.txt:
Go update based on UG. useless if its not updated for PE

Ui:
``` Suggestion
    public void printTrackerDiagram(ModuleTrackerList moduleTrackerList) {
        printLine();
        for (int i = 0; i < moduleTrackerList.getSize(); i++) {
            ModuleTracker moduleTracker = moduleTrackerList.getModuleTracker(i);
            int timeTaken = (int) moduleTracker.getTotalTimeTaken();
            int segmentedTimeTaken = timeTaken / TEN_MINUTES_BLOCK;
            for (int j = 0; j < segmentedTimeTaken; j++) {
                if (j == FIRST_VAR) {
                    System.out.println("| ");
                    System.out.print("| #");
                } else {
                    System.out.print("#");
                }
            }
            String moduleCode = moduleTracker.getModuleCode();
            System.out.print("\t" + moduleCode + " -- " + timeTaken + " minutes\n");
        }
    }
```
Nested loop, not enough abstraction. can do better at least for inner for loop.

StartTrackerCommand:
``` Suggestion
    public void execute(SemesterList semesterList, TaskList tasks, Ui ui, Storage storage) throws OofException {
        if (description.isEmpty()) {
            throw new OofException("Please enter a Task!");
            throw new OofException("Please enter the Assignment module code and description!");
        }
        Task task = findTask(tasks, description);
        if (task.getStatus()) {
            throw new OofException("Task has already been completed.");

        String[] input = description.split(" ", 2);
        if (input.length < MINIMUM_SIZE) {
            throw new OofException("Invalid input!");
        }
```
whats the 2. same in stoptrackercommand.

**Comment by debbiextan** [comment]
Thanks, @KahLokKee and @jasperosy for your reviews and suggestions. The code has been refactored alongside the UG and DG updated accordingly.

**Review by jasperosy — COMMENTED** [review]
UserGuide:
mentioned in previous comments that images need to be labelled. Follow the labelling convention. 
Also, can you follow the tone of the previous features? You are not selling your feature but explaining your feature with you language. 
Follow example format, language, label images
Please follow format for consistency.

Manual:
i dont see any correct changes in your manual... Can you take a look at the command summary again and edit your manual accordingly
e.g. no sort command, recurring wrong, viewweek wrong, calendar wrong, reminder also wrong, missing out commands....

CommandParser:
Also, why are you tampering with my methods for CommandParser?? It was originally correct and now you deleted and pasted it back.

**Comment by debbiextan** [comment]
@jasperosy 
The CommandParser edits were created unintentionally during a merge conflict fix. I've replaced all the methods manually from the latest master copy of CommandParser. Do take a look and double-check that your methods back to their original.

UG label images have been edited to fit our standardized format and edited tone for tracker related features. If tone of Semester features need editing, @Kenlhc will need to make his necessary edits as well.

Thanks for pointing out the manual.txt issue, it has been rectified according to the available commands we have noted down in the UG.

**Comment by jasperosy** [comment]
> @jasperosy
> The CommandParser edits were created unintentionally during a merge conflict fix. I've replaced all the methods manually from the latest master copy of CommandParser. Do take a look and double-check that your methods back to their original.
> 
> UG label images have been edited to fit our standardized format and edited tone for tracker related features. If tone of Semester features need editing, @Kenlhc will need to make his necessary edits as well.
> 
> Thanks for pointing out the manual.txt issue, it has been rectified according to the available commands we have noted down in the UG.

That brings about another point. Have you ensured that your UG is up to date with the copy in the repo now? Because I have already rectified @Kenlhc part.

**Review by kahlokkee — APPROVED** [review]
